File: flutter/lib/screens/buttons_screen.dart
import 'package:flutter/material.dart';
import '../state/catalog_state.dart';
import '../widgets/catalog_card.dart';

class ButtonsScreen extends StatefulWidget {
  final CatalogState state;

  const ButtonsScreen({super.key, required this.state});

  @override
  State<ButtonsScreen> createState() => _ButtonsScreenState();
}

class _ButtonsScreenState extends State<ButtonsScreen> {
  String _feedbackMessage = 'Respuesta: Pulsa cualquier botón para interactuar';
  Set<int> _selectedSegment = {0};
  bool _isLoading = false;

  void _triggerFeedback(String buttonName) {
    final msg = 'Respuesta visible: Pulsado "$buttonName"';
    setState(() {
      _feedbackMessage = msg;
    });
    widget.state.setLastAction(msg);
    ScaffoldMessenger.of(context).clearSnackBars();
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text(msg),
        duration: const Duration(seconds: 1),
        behavior: SnackBarBehavior.floating,
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);

    return SingleChildScrollView(
      padding: const EdgeInsets.all(16.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            'Sección 2: Botones y Acciones',
            style: theme.textTheme.headlineMedium?.copyWith(
              color: theme.colorScheme.primary,
              fontWeight: FontWeight.bold,
            ),
          ),
          Text(
            'Componentes interactivos para disparar acciones con feedback inmediato visual y háptico.',
            style: theme.textTheme.bodyMedium?.copyWith(
              color: theme.colorScheme.onSurfaceVariant,
            ),
          ),
          const SizedBox(height: 12),

          // Banner de retroalimentación inmediata
          Card(
            color: theme.colorScheme.surfaceContainerHighest,
            child: Padding(
              padding: const EdgeInsets.all(14.0),
              child: Row(
                children: [
                  Icon(Icons.smart_button, color: theme.colorScheme.primary),
                  const SizedBox(width: 12),
                  Expanded(
                    child: Text(
                      _feedbackMessage,
                      style: theme.textTheme.labelLarge?.copyWith(
                        color: theme.colorScheme.primary,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ),

          const SizedBox(height: 8),

          // 1. Jerarquía de botones
          CatalogCard(
            title: '1. Jerarquía de Énfasis: Relleno, Contorno y Solo Texto',
            description: 'Permiten jerarquizar acciones en la vista según su relevancia: primaria (FilledButton), secundaria (OutlinedButton) y terciaria (TextButton).',
            child: Row(
              children: [
                Expanded(
                  child: FilledButton(
                    onPressed: () => _triggerFeedback('FilledButton (Relleno)'),
                    child: const Text('Relleno'),
                  ),
                ),
                const SizedBox(width: 8),
                Expanded(
                  child: OutlinedButton(
                    onPressed: () => _triggerFeedback('OutlinedButton (Contorno)'),
                    child: const Text('Contorno'),
                  ),
                ),
                const SizedBox(width: 8),
                Expanded(
                  child: TextButton(
                    onPressed: () => _triggerFeedback('TextButton (Texto)'),
                    child: const Text('Texto'),
                  ),
                ),
              ],
            ),
          ),

          // 2. Botones con ícono
          CatalogCard(
            title: '2. Botones con Ícono (Solo Ícono y Mixto)',
            description: 'Comunican la acción mediante iconos universales, ya sea de forma compacta (IconButton) o acompañada de texto explicativo.',
            child: Row(
              children: [
                IconButton.filledTonal(
                  onPressed: () => _triggerFeedback('IconButton (Favorito)'),
                  icon: const Icon(Icons.favorite),
                ),
                const SizedBox(width: 16),
                Expanded(
                  child: FilledButton.icon(
                    onPressed: () => _triggerFeedback('FilledButton con Ícono'),
                    icon: const Icon(Icons.favorite),
                    label: const Text('Favorito con Texto'),
                  ),
                ),
              ],
            ),
          ),

          // 3. Botones de Acción Flotante (FAB)
          CatalogCard(
            title: '3. Botón de Acción Flotante (FAB Estándar y Extendido)',
            description: 'Se eleva por encima del contenido para promover la acción principal. Su variante extendida incluye etiqueta de texto.',
            child: Row(
              children: [
                FloatingActionButton(
                  heroTag: 'fab_flutter_normal',
                  onPressed: () => _triggerFeedback('FAB Normal Circular'),
                  child: const Icon(Icons.add),
                ),
                const SizedBox(width: 20),
                FloatingActionButton.extended(
                  heroTag: 'fab_flutter_extended',
                  onPressed: () => _triggerFeedback('FloatingActionButton.extended'),
                  icon: const Icon(Icons.add),
                  label: const Text('Nuevo Registro'),
                ),
              ],
            ),
          ),

          // 4. Selector segmentado
          CatalogCard(
            title: '4. Selector Segmentado (SegmentedButton)',
            description: 'Agrupa un conjunto de opciones mutuamente excluyentes en una franja horizontal continua bajo el estándar Material 3.',
            child: SizedBox(
              width: double.infinity,
              child: SegmentedButton<int>(
                segments: const [
                  ButtonSegment(value: 0, label: Text('Día')),
                  ButtonSegment(value: 1, label: Text('Semana')),
                  ButtonSegment(value: 2, label: Text('Mes')),
                ],
                selected: _selectedSegment,
                onSelectionChanged: (Set<int> newSelection) {
                  setState(() {
                    _selectedSegment = newSelection;
                  });
                  final labels = ['Día', 'Semana', 'Mes'];
                  _triggerFeedback('Segmento: ${labels[newSelection.first]}');
                },
              ),
            ),
          ),

          // 5. Botón deshabilitado y botón en estado de carga
          CatalogCard(
            title: '5. Botón Deshabilitado y Botón con Estado de Carga',
            description: 'Comunican la falta de disponibilidad de la acción (onPressed: null) o la ejecución activa de una tarea asíncrona mediante un spinner.',
            child: Row(
              children: [
                const Expanded(
                  child: FilledButton(
                    onPressed: null,
                    child: Text('Deshabilitado'),
                  ),
                ),
                const SizedBox(width: 12),
                Expanded(
                  child: FilledButton.tonal(
                    onPressed: _isLoading
                        ? null
                        : () async {
                            setState(() {
                              _isLoading = true;
                              _feedbackMessage = 'Iniciando proceso asíncrono simulado...';
                            });
                            await Future.delayed(const Duration(seconds: 2));
                            if (mounted) {
                              setState(() {
                                _isLoading = false;
                              });
                              _triggerFeedback('Carga asíncrona completada con éxito');
                            }
                          },
                    child: _isLoading
                        ? const SizedBox(
                            width: 18,
                            height: 18,
                            child: CircularProgressIndicator(strokeWidth: 2),
                          )
                        : const Text('Iniciar Carga'),
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

File: android-compose/app/src/main/java/com/escom/tarea2/compose/ui/screens/ButtonsScreen.kt
package com.escom.tarea2.compose.ui.screens

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.filled.SmartButton
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExtendedFloatingActionButton
import androidx.compose.material3.FilledTonalButton
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.escom.tarea2.compose.ui.components.CatalogItemCard
import com.escom.tarea2.compose.viewmodel.SharedCatalogViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ButtonsScreen(viewModel: SharedCatalogViewModel) {
    var buttonFeedback by remember { mutableStateOf("Respuesta: Pulsa cualquier botón para interactuar") }
    var selectedSegmentIndex by remember { mutableIntStateOf(0) }
    val segmentOptions = listOf("Día", "Semana", "Mes")

    var isLoading by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()

    fun triggerFeedback(name: String) {
        val msg = "Respuesta visible: Pulsado '$name'"
        buttonFeedback = msg
        viewModel.setLastAction(msg)
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(16.dp)
    ) {
        Text(
            text = "Sección 2: Botones y Acciones",
            style = MaterialTheme.typography.headlineMedium,
            color = MaterialTheme.colorScheme.primary
        )
        Text(
            text = "Disparadores de eventos táctiles. Todos los botones producen una respuesta visible y tangible inmediata.",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        Spacer(modifier = Modifier.height(12.dp))

        // Banner de respuesta visual inmediata
        Card(
            modifier = Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(10.dp),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant)
        ) {
            Row(
                modifier = Modifier.padding(14.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Icon(
                    imageVector = Icons.Default.SmartButton,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.primary
                )
                Spacer(modifier = Modifier.width(12.dp))
                Text(
                    text = buttonFeedback,
                    style = MaterialTheme.typography.labelLarge,
                    color = MaterialTheme.colorScheme.primary
                )
            }
        }

        Spacer(modifier = Modifier.height(8.dp))

        // 1. Jerarquía de botones
        CatalogItemCard(
            title = "1. Jerarquía de Énfasis: Relleno, Contorno y Solo Texto",
            description = "Estructuran las llamadas a la acción según su prioridad en pantalla: acción primaria destacada, secundaria perimetral y terciaria ligera."
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Button(
                    onClick = { triggerFeedback("Button Relleno (Primary)") },
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Relleno")
                }
                OutlinedButton(
                    onClick = { triggerFeedback("OutlinedButton (Secondary)") },
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Contorno")
                }
                TextButton(
                    onClick = { triggerFeedback("TextButton (Tertiary)") },
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Texto")
                }
            }
        }

        // 2. Botones con ícono
        CatalogItemCard(
            title = "2. Botones con Ícono (Solo Ícono y Mixto)",
            description = "Sintetizan la intención de la interacción mediante glifos vectoriales que refuerzan el reconocimiento semántico inmediato."
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically
            ) {
                IconButton(
                    onClick = { triggerFeedback("IconButton de Favorito") }
                ) {
                    Icon(
                        imageVector = Icons.Default.Favorite,
                        contentDescription = "Favorito",
                        tint = MaterialTheme.colorScheme.primary
                    )
                }

                Spacer(modifier = Modifier.width(16.dp))

                Button(
                    onClick = { triggerFeedback("Botón con Ícono y Texto") },
                    modifier = Modifier.weight(1f)
                ) {
                    Icon(Icons.Default.Favorite, contentDescription = null)
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Añadir a Favoritos")
                }
            }
        }

        // 3. Botones de Acción Flotante (FAB)
        CatalogItemCard(
            title = "3. Botón de Acción Flotante (Normal y Extendido)",
            description = "Componente flotante sobre la superficie para la acción prioritaria de la pantalla, con variante extendida que añade texto explicativo."
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically
            ) {
                FloatingActionButton(
                    onClick = { triggerFeedback("FAB Normal Circular") }
                ) {
                    Icon(Icons.Default.Add, contentDescription = "Agregar")
                }

                Spacer(modifier = Modifier.width(16.dp))

                ExtendedFloatingActionButton(
                    onClick = { triggerFeedback("Extended FAB con Rótulo") },
                    icon = { Icon(Icons.Default.Add, contentDescription = null) },
                    text = { Text("Crear Registro") }
                )
            }
        }

        // 4. Selector Segmentado (SegmentedButton)
        CatalogItemCard(
            title = "4. Selector Segmentado (SingleChoiceSegmentedButtonRow)",
            description = "Agrupación horizontal de opciones contiguas mutuamente excluyentes implementada con el estándar de Material 3."
        ) {
            SingleChoiceSegmentedButtonRow(modifier = Modifier.fillMaxWidth()) {
                segmentOptions.forEachIndexed { index, label ->
                    SegmentedButton(
                        shape = SegmentedButtonDefaults.itemShape(index = index, count = segmentOptions.size),
                        onClick = {
                            selectedSegmentIndex = index
                            triggerFeedback("Segmento seleccionado: $label")
                        },
                        selected = index == selectedSegmentIndex
                    ) {
                        Text(label)
                    }
                }
            }
        }

        // 5. Botón deshabilitado y botón con estado de carga
        CatalogItemCard(
            title = "5. Botón Deshabilitado y Estado de Carga (Loading)",
            description = "Representa estados de bloqueo temporal ante validaciones insatisfechas o ejecución asíncrona mediante un indicador de giro circular."
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Button(
                    onClick = {},
                    enabled = false,
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Inactivo")
                }

                FilledTonalButton(
                    onClick = {
                        if (!isLoading) {
                            isLoading = true
                            buttonFeedback = "Ejecutando tarea asíncrona..."
                            scope.launch {
                                delay(2000)
                                isLoading = false
                                triggerFeedback("Operación asíncrona completada exitosamente")
                            }
                        }
                    },
                    modifier = Modifier.weight(1f)
                ) {
                    if (isLoading) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(18.dp),
                            strokeWidth = 2.dp
                        )
                        Spacer(modifier = Modifier.width(8.dp))
                        Text("Cargando...")
                    } else {
                        Text("Iniciar Carga")
                    }
                }
            }
        }
    }
}
